sim/src/cap.rs

```diff
@@ -0,0 +1,87 @@
+use crate::{CarID, VehicleType};
+use geom::{Duration, Time};
+use map_model::{LaneID, Map, Path, PathConstraints, PathStep};
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, BTreeSet};
+
+// Note this only indexes into the zones we track here, not all of them in the map.
+type ZoneIdx = usize;
+
+// This only caps driving trips.
+#[derive(Serialize, Deserialize, PartialEq, Clone)]
+pub struct CapSimState {
+    lane_to_zone: BTreeMap<LaneID, ZoneIdx>,
+    zones: Vec<Zone>,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Clone)]
+struct Zone {
+    cap: usize,
+    entered_in_last_hour: BTreeSet<CarID>,
+    // TODO Maybe want sliding windows or something else
+    hour_started: Time,
+}
+
+impl CapSimState {
+    pub fn new(map: &Map) -> CapSimState {
+        let mut sim = CapSimState {
+            lane_to_zone: BTreeMap::new(),
+            zones: Vec::new(),
+        };
+        for z in map.all_zones() {
+            if let Some(cap) = z.restrictions.cap_vehicles_per_hour {
+                let idx = sim.zones.len();
+                for r in &z.members {
+                    for l in map.get_r(*r).all_lanes() {
+                        if PathConstraints::Car.can_use(map.get_l(l), map) {
+                            sim.lane_to_zone.insert(l, idx);
+                        }
+                    }
+                }
+                sim.zones.push(Zone {
+                    cap,
+                    entered_in_last_hour: BTreeSet::new(),
+                    hour_started: Time::START_OF_DAY,
+                });
+            }
+        }
+        sim
+    }
+
+    pub fn car_entering_lane(&mut self, now: Time, car: CarID, lane: LaneID) {
+        if car.1 != VehicleType::Car {
+            return;
+        }
+        let zone = if let Some(idx) = self.lane_to_zone.get(&lane) {
+            &mut self.zones[*idx]
+        } else {
+            return;
+        };
+
+        if now - zone.hour_started >= Duration::hours(1) {
+            zone.hour_started = Time::START_OF_DAY + Duration::hours(now.get_parts().0);
+            zone.entered_in_last_hour.clear();
+        }
+        zone.entered_in_last_hour.insert(car);
```

Comment:
Say we have an hourly cap of 1000.

car 1-999 enters at 12:59 AM.

2 mins later
car 1000-1998 enters at 1:02 AM

Even though almost 2k cars have entered in 3 minutes, at that point we're still "under the cap" because of how we're bucketing (vs the "sliding window" that you mentioned). I think this is fine, just making sure I'm understanding.

Reply:
Correct. Now that there's a prototype, I'll ping the people interested in this and see if sliding windows make more sense.

Another weird problem: Say the cap is 500. At the start of the hour, 1000 cars all appear from the border and each check if the cap has been exceeded. It hasn't, so they all go. Only once they physically enter the zone does the count increase, so a car spawning after that would see. Might want to reserve the slots as soon as the car spawns. The slightly weird thing is if the car makes the reservation from far away and enters the zone much much later.